### PR TITLE
Feature/query params cleanup

### PR DESCRIPTION
**Pull request by mohitb35**
1. Creates a template interface QueryParamContextInterface to model QueryParamContext
2. Adds types for project, user, donation, utility types and updates common types.
3. Adds suitable type guards/annotations at various places.

> This does not resolve every TS warning in the project. It focuses on the warnings shown for `QueryParamContext` and updates code related to the changes to `QueryParamContext` and `QueryParamContextInterface`.

### CODE REVIEW

**Comment by mariahosfeld** [comment]
When making a test donation via PayPal, in some cases I can see the button for a split second, after that it is missing
grafik

**Comment by mariahosfeld** [comment]
Functionality now works for me.

File: src/Donations/Components/DonationsForm.tsx [hunk 1]
```diff
@@ -216,78 +224,77 @@ function DonationsForm() {
       default:
         paymentLabel = t("treesInCountry", {
           treeCount: quantity,
-          country: t(`country:${projectDetails.country?.toLowerCase()}`),
+          // country: t(`country:${projectDetails.country?.toLowerCase()}`),
```
Comment (sunilsabatp) [inline src/Donations/Components/DonationsForm.tsx:227]: if these line of code not needed then remove it from here

File: src/Donations/Components/PaymentsForm.tsx [hunk 2]
```diff
@@ -93,6 +93,10 @@ function PaymentsForm({}: Props): ReactElement {
     method: string,
     providerObject?: any
   ) => {
+    if (!paymentSetup) {
+      console.log("Missing payment options");
+      return;
+    }
```
Comment (sunilsabatp) [inline src/Donations/Components/PaymentsForm.tsx:99]: is this log is necessary ?

Comment (mohitb35) [reply src/Donations/Components/PaymentsForm.tsx:99]: Yes, in the unlikely case that `paymentSetup` is not initialized, we'd be able to diagnose the problem quickly. At the same time, this isn't a message that needs to be shown to the user.

This scenario is unlikely to occur, so this log will be probably never be shown.

File: src/Donations/Components/DonationsForm.tsx (hunk 1 above)
Comment (mohitb35) [reply src/Donations/Components/DonationsForm.tsx]: Done in - 2b80002f15975bc7214694c3593e86ec6047643f

**Review by sunilsabatp — APPROVED** [review]

**Pull request merged**